File: server/src/services/ai.service.ts
import { GoogleGenAI } from '@google/genai';
import OpenAI from 'openai';

const geminiApiKey = process.env.GEMINI_API_KEY || '';
const openaiApiKey = process.env.OPENAI_API_KEY || '';

let aiProvider: 'gemini' | 'openai' | 'mock' = 'mock';

if (geminiApiKey) {
  aiProvider = 'gemini';
} else if (openaiApiKey) {
  aiProvider = 'openai';
}

console.log(`🤖 Kintsu AI Integration Provider Mode: [${aiProvider.toUpperCase()}]`);

const googleAI = geminiApiKey ? new GoogleGenAI({ apiKey: geminiApiKey }) : null;
const openai = openaiApiKey ? new OpenAI({ apiKey: openaiApiKey }) : null;

export const getAIStatus = () => ({
  provider: aiProvider,
  hasGeminiKey: Boolean(geminiApiKey),
  hasOpenAIKey: Boolean(openaiApiKey),
  supportedFeatures: [
    'Rehabilitation Session Builder',
    'Participant Behavioral & Empathy Analysis',
    'Motivational Story Generator',
    'Book Insight & Discussion Prompt Generator'
  ]
});

/**
 * Low-level text completion helper that uses Gemini or OpenAI based on available API key
 */
async function generateText(prompt: string, systemInstruction?: string): Promise<string> {
  if (aiProvider === 'gemini' && googleAI) {
    try {
      const response = await googleAI.models.generateContent({
        model: 'gemini-2.5-flash',
        contents: prompt,
        config: systemInstruction ? { systemInstruction } : undefined
      });
      return response.text || '';
    } catch (err) {
      console.warn('Gemini API call failed, falling back to mock generator:', err);
    }
  }

  if (aiProvider === 'openai' && openai) {
    try {
      const response = await openai.chat.completions.create({
        model: 'gpt-4o-mini',
        messages: [
          ...(systemInstruction ? [{ role: 'system' as const, content: systemInstruction }] : []),
          { role: 'user' as const, content: prompt }
        ]
      });
      return response.choices[0]?.message?.content || '';
    } catch (err) {
      console.warn('OpenAI API call failed, falling back to mock generator:', err);
    }
  }

  // Fallback mock response if no API keys configured or call failed
  return '';
}

/**
 * Generate a complete structured rehabilitation classroom session
 */
export async function generateSessionAI(topic: string, category: string, block?: string) {
  const systemInstruction = `You are an expert prison rehabilitation counselor and psychologist designing structured classroom sessions for inmate transformation. Output clean JSON only.`;

  const prompt = `Design a 6-step rehabilitation classroom session on the topic "${topic}" in category "${category}" for block "${block || 'Block C'}".
Return a valid JSON object with the following structure:
{
  "title": "string",
  "subtitle": "string",
  "category": "${category}",
  "description": "string",
  "steps": [
    { "stepNumber": 1, "type": "video", "title": "string", "durationMinutes": 7, "content": "string" },
    { "stepNumber": 2, "type": "story", "title": "string", "durationMinutes": 10, "content": "string" },
    { "stepNumber": 3, "type": "discussion", "title": "string", "durationMinutes": 15, "content": "string", "questions": ["q1", "q2", "q3"] },
    { "stepNumber": 4, "type": "activity", "title": "string", "durationMinutes": 10, "content": "string" },
    { "stepNumber": 5, "type": "reflection", "title": "string", "durationMinutes": 8, "content": "string" },
    { "stepNumber": 6, "type": "closing", "title": "string", "durationMinutes": 5, "content": "string" }
  ]
}`;

  const rawText = await generateText(prompt, systemInstruction);

  if (rawText) {
    try {
      // Extract json block if present
      const jsonMatch = rawText.match(/\{[\s\S]*\}/);
      if (jsonMatch) {
        return JSON.parse(jsonMatch[0]);
      }
    } catch (e) {
      console.error('Failed to parse AI session response as JSON, returning fallback structure');
    }
  }

  // Smart Fallback Generation
  return {
    title: `${topic}: Mindset & Emotional Growth`,
    subtitle: `Structured Group Learning Session on ${category}`,
    category,
    description: `A comprehensive rehabilitation session focusing on ${topic.toLowerCase()}, designed to promote self-reflection and impulse control.`,
    steps: [
      { stepNumber: 1, type: 'video', title: `Introductory Video: Understanding ${topic}`, durationMinutes: 7, content: `Visual presentation introducing core principles of ${topic.toLowerCase()}.` },
      { stepNumber: 2, type: 'story', title: `Real-Life Transformation Story`, durationMinutes: 10, content: `A narrative of an individual who mastered ${topic.toLowerCase()} after overcoming major life obstacles.` },
      { stepNumber: 3, type: 'discussion', title: `Instructor-Led Discussion`, durationMinutes: 15, content: `Interactive dialogue exploring real-world applications of ${topic.toLowerCase()}.`, questions: [`How does ${topic.toLowerCase()} affect your daily choices?`, `What is one trigger you can control starting today?`] },
      { stepNumber: 4, type: 'activity', title: `Group Guided Exercise`, durationMinutes: 10, content: `Practical group exercise reinforcing positive behavioral choices.` },
      { stepNumber: 5, type: 'reflection', typeName: 'reflection', title: `Personal Reflection & Journaling`, durationMinutes: 8, content: `Writing down one personal commitment for the week.` },
      { stepNumber: 6, type: 'closing', title: `Closing Summary by Case Worker`, durationMinutes: 5, content: `Encouraging closing message highlighting personal responsibility and growth.` }
    ]
  };
}

/**
 * Analyze participant behavior and generate empathy & de-escalation insights
 */
export async function analyzeParticipantAI(participantName: string, caseNotes: string[], recentResponse?: string) {
  const systemInstruction = `You are a forensic psychologist and correctional caseworker analyzing participant progress. Return clean JSON only.`;

  const prompt = `Analyze participant "${participantName}" based on notes: ${JSON.stringify(caseNotes)} and recent response: "${recentResponse || 'Active classroom participant'}".
Return JSON:
{
  "deEscalationScore": 82,
  "empathyScore": 76,
  "summaryAnalysis": "string",
  "keyStrengths": ["s1", "s2"],
  "areasForGrowth": ["g1", "g2"],
  "recommendedModule": "string"
}`;

  const rawText = await generateText(prompt, systemInstruction);

  if (rawText) {
    try {
      const jsonMatch = rawText.match(/\{[\s\S]*\}/);
      if (jsonMatch) return JSON.parse(jsonMatch[0]);
    } catch (e) {}
  }

  return {
    deEscalationScore: 82,
    empathyScore: 78,
    summaryAnalysis: `${participantName} demonstrates steady engagement and emotional awareness during classroom sessions, with positive progress in de-escalation techniques.`,
    keyStrengths: ['Active listening in group discussions', 'Willingness to acknowledge personal triggers'],
    areasForGrowth: ['Refining impulse control under direct verbal confrontation'],
    recommendedModule: '5-4-3-2-1 Calm Routine & Verbal De-escalation'
  };
}

/**
 * Generate a inspirational transformation story for classroom reading
 */
export async function generateStoryAI(theme: string, targetAudience: string = 'Inmate Rehabilitation Group') {
  const systemInstruction = `You write uplifting, realistic, non-preachy transformation stories for prison classroom rehabilitation. Output JSON only.`;

  const prompt = `Write an inspiring rehabilitation story about "${theme}" for ${targetAudience}.
Return JSON:
{
  "title": "string",
  "category": "${theme}",
  "summary": "string",
  "fullText": "string",
  "moral": "string",
  "discussionQuestions": ["q1", "q2"]
}`;

  const rawText = await generateText(prompt, systemInstruction);

  if (rawText) {
    try {
      const jsonMatch = rawText.match(/\{[\s\S]*\}/);
      if (jsonMatch) return JSON.parse(jsonMatch[0]);
    } catch (e) {}
  }

  return {
    title: `The Turning Point of Perseverance: ${theme}`,
    category: theme,
    summary: `An inspiring tale of choosing self-discipline and accountability over reaction.`,
    fullText: `Change does not happen in a single grand event; it happens in quiet, unglamorous moments when a person decides to act differently than they did yesterday. When faced with frustration, choosing patience over impulse is where freedom truly begins...`,
    moral: `Your past explains where you have been, but it never defines where you can go.`,
    discussionQuestions: [
      `What choice in this story resonated most with you?`,
      `How can you apply this lesson when facing frustration this week?`
    ]
  };
}

File: server/services/ai_service.py
import json
import re
from google import genai
from config import GEMINI_API_KEY

ai_provider = "gemini" if GEMINI_API_KEY else "mock"
print(f"[Kintsu FastAPI AI Integration] Provider Mode: [{ai_provider.upper()}]")

client = genai.Client(api_key=GEMINI_API_KEY) if GEMINI_API_KEY else None

def get_ai_status():
    return {
        "provider": "gemini",
        "activeMode": ai_provider,
        "hasGeminiKey": bool(GEMINI_API_KEY),
        "model": "gemini-2.5-flash",
        "framework": "FastAPI + google-genai",
        "supportedFeatures": [
            "Rehabilitation Session Builder",
            "Participant Behavioral & Empathy Analysis",
            "Motivational Story Generator",
            "Roleplay Simulation Dialogue Engine"
        ]
    }

def generate_text(prompt: str, system_instruction: str = "") -> str:
    if ai_provider == "gemini" and client:
        try:
            config = {"system_instruction": system_instruction} if system_instruction else None
            response = client.models.generate_content(
                model="gemini-2.5-flash",
                contents=prompt,
                config=config
            )
            return response.text or ""
        except Exception as e:
            print(f"Warning: Gemini API call failed, using mock generator: {e}")
    return ""

def generate_session_ai(topic: str, category: str, block: str = "Block C"):
    system_instruction = "You are an expert prison rehabilitation counselor and forensic psychologist designing structured classroom sessions for inmate mindset transformation. Output clean JSON only."
    prompt = f"""Design a 6-step rehabilitation classroom session on the topic "{topic}" in category "{category}" for block "{block}".
Return a valid JSON object with the following structure:
{{
  "title": "string",
  "subtitle": "string",
  "category": "{category}",
  "description": "string",
  "steps": [
    {{ "stepNumber": 1, "type": "video", "title": "string", "durationMinutes": 7, "content": "string" }},
    {{ "stepNumber": 2, "type": "story", "title": "string", "durationMinutes": 10, "content": "string" }},
    {{ "stepNumber": 3, "type": "discussion", "title": "string", "durationMinutes": 15, "content": "string", "questions": ["q1", "q2", "q3"] }},
    {{ "stepNumber": 4, "type": "activity", "title": "string", "durationMinutes": 10, "content": "string" }},
    {{ "stepNumber": 5, "type": "reflection", "title": "string", "durationMinutes": 8, "content": "string" }},
    {{ "stepNumber": 6, "type": "closing", "title": "string", "durationMinutes": 5, "content": "string" }}
  ]
}}"""

    raw_text = generate_text(prompt, system_instruction)
    if raw_text:
        try:
            match = re.search(r'\{[\s\S]*\}', raw_text)
            if match:
                return json.loads(match.group(0))
        except Exception as e:
            print(f"Error parsing JSON from Gemini: {e}")

    # Fallback structure
    return {
        "title": f"{topic}: Mindset & Emotional Growth",
        "subtitle": f"Structured Group Learning Session on {category}",
        "category": category,
        "description": f"A comprehensive rehabilitation session focusing on {topic.lower()}, guided by Gemini AI.",
        "steps": [
            { "stepNumber": 1, "type": "video", "title": f"Introductory Video: Understanding {topic}", "durationMinutes": 7, "content": f"Visual presentation introducing core principles of {topic.lower()}." },
            { "stepNumber": 2, "type": "story", "title": "Real-Life Transformation Story", "durationMinutes": 10, "content": f"A narrative of an individual who mastered {topic.lower()} after overcoming major life obstacles." },
            { "stepNumber": 3, "type": "discussion", "title": "Instructor-Led Discussion", "durationMinutes": 15, "content": f"Interactive dialogue exploring real-world applications of {topic.lower()}.", "questions": [f"How does {topic.lower()} affect your daily choices?", "What is one trigger you can control starting today?"] },
            { "stepNumber": 4, "type": "activity", "title": "Group Guided Exercise", "durationMinutes": 10, "content": "Practical group exercise reinforcing positive behavioral choices." },
            { "stepNumber": 5, "type": "reflection", "title": "Personal Reflection & Journaling", "durationMinutes": 8, "content": "Writing down one personal commitment for the week." },
            { "stepNumber": 6, "type": "closing", "title": "Closing Summary by Case Worker", "durationMinutes": 5, "content": "Encouraging closing message highlighting personal responsibility and growth." }
        ]
    }

def analyze_participant_ai(participant_name: str, case_notes: list, recent_response: str = None):
    system_instruction = "You are a forensic psychologist analyzing participant progress. Return clean JSON only."
    prompt = f"""Analyze participant "{participant_name}" based on notes: {json.dumps(case_notes)} and recent response: "{recent_response or 'Active participant'}".
Return JSON:
{{
  "deEscalationScore": 84,
  "empathyScore": 78,
  "summaryAnalysis": "string",
  "keyStrengths": ["s1", "s2"],
  "areasForGrowth": ["g1", "g2"],
  "recommendedModule": "string"
}}"""

    raw_text = generate_text(prompt, system_instruction)
    if raw_text:
        try:
            match = re.search(r'\{[\s\S]*\}', raw_text)
            if match:
                return json.loads(match.group(0))
        except Exception:
            pass

    return {
        "deEscalationScore": 84,
        "empathyScore": 78,
        "summaryAnalysis": f"{participant_name} demonstrates steady engagement and emotional awareness during classroom sessions.",
        "keyStrengths": ["Active listening in group discussions", "Willingness to acknowledge personal triggers"],
        "areasForGrowth": ["Refining impulse control under direct verbal confrontation"],
        "recommendedModule": "5-4-3-2-1 Calm Routine & Verbal De-escalation"
    }
